### PR TITLE
Don't try to write the wrong UID, version on Federated Ingress updates.

### LINKED ISSUE #33135: Fix federation ingress e2e tests
They are failing right now due to a panic which is being fixed by https://github.com/kubernetes/kubernetes/pull/33120.

They still fail after that fix.
I see the following in the controller manager logs:
Logs: https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/33120/kubernetes-pull-build-test-federation-e2e-gce/30/build-log.txt
```
ingress_controller.go:748] Failed to execute updates for e2e-tests-federated-ingress-jo8tc/federated-ingress: Operation cannot be fulfilled on ingresses.extensions "federated-ingress": StorageError: invalid object, Code: 4, Key: /registry/ingress/e2e-tests-federated-ingress-jo8tc/federated-ingress, ResourceVersion: 0, AdditionalErrorMsg: Precondition failed: UID in precondition: 0xc82529be60, UID in object meta: b465807f-7f78-11e6-b65f-42010af00002
```

Update fails because of UID mismatch

### PR DESCRIPTION
**Pull request by ghost**
Fixes #33135.

This looks more complicated than it really is. 
Essentially, use the cluster object's metadata, rather than the federated objects's metadata when updating cluster Ingress objects.  The deepcopy stuff is mainly to get around shortcomings in the Kubernetes fake test infrastructure, which ends up with crossed pointers if we don't deep copy.

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/33339)

### CODE REVIEW

**Comment by quinton-hoole** [comment]
@k8s-bot Federation gce e2e test this

**Comment by quinton-hoole** [comment]
@k8s-bot federation gce e2e test this

**Comment by quinton-hoole** [comment]
@k8s-bot federation gce e2e test this

**Comment by quinton-hoole** [comment]
This passes unit test locally, and is a definite fix to a definite bug.  I have not been able to run Federated e2e tests locally yet, so it's possible that those will still fail on Jenkins.  Either way it's worth merging this PR into head.  It's probably a good idea to hold off cherrypicking into the v1.4 branch until we've succeeded in getting the Federated e2e tests to pass.

**Comment by nikhiljindal** [comment]
Cluster failed to come up in federation e2e.
I dont see an obvious reason.
Retrying:
@k8s-bot federation gce e2e test this

File: federation/pkg/federation-controller/util/meta.go [hunk 1]
```diff
@@ -34,8 +34,28 @@ func CopyObjectMeta(obj api_v1.ObjectMeta) api_v1.ObjectMeta {
 	}
 }
 
+// Deep copies cluster-independent, user provided data from the given ObjectMeta struct. If in
+// the future the ObjectMeta structure is expanded then any field that is not populated
+// by the api server should be included here.
+func DeepCopyObjectMeta(obj api_v1.ObjectMeta) api_v1.ObjectMeta {
+	copyMeta := CopyObjectMeta(obj)
+	if obj.Labels != nil {
```
Comment (nikhiljindal) [inline federation/pkg/federation-controller/util/meta.go]: Add a comment regarding why is this required?
CopyObjectMata should set the right labels and annotations

File: federation/pkg/federation-controller/ingress/ingress_controller_test.go [hunk 2]
```diff
@@ -96,10 +96,10 @@ func TestIngressController(t *testing.T) {
...
 	ing1 := extensions_v1beta1.Ingress{
 		ObjectMeta: api_v1.ObjectMeta{
-			Name:        "test-ingress",
-			Namespace:   "mynamespace",
-			SelfLink:    "/api/v1/namespaces/mynamespace/ingress/test-ingress",
-			Annotations: map[string]string{},
+			Name:      "test-ingress",
+			Namespace: "mynamespace",
+			SelfLink:  "/api/v1/namespaces/mynamespace/ingress/test-ingress",
+			// TODO: Remove: Annotations: map[string]string{},
```
Comment (nikhiljindal) [inline federation/pkg/federation-controller/ingress/ingress_controller_test.go]: remove?

File: federation/pkg/federation-controller/ingress/ingress_controller.go [hunk 3]
```diff
@@ -707,29 +703,43 @@ func (ic *IngressController) reconcileIngress(ingress types.NamespacedName) {
...
+				}
+				ic.eventRecorder.Eventf(baseIngress, api.EventTypeNormal, "UpdateInCluster",
+					"Updating ingress in cluster %s", cluster.Name)
+
+				operations = append(operations, util.FederatedOperation{
+					Type:        util.OperationTypeUpdate,
+					Obj:         desiredIngress,
+					ClusterName: cluster.Name,
+				})
+				// TODO: Transfer any readonly (target-proxy, url-map etc) annotations from the master cluster to the federation, if this is the master cluster.
```
Comment (nikhiljindal) [inline federation/pkg/federation-controller/ingress/ingress_controller.go]: Sorry did not understand why do you want to transfer annotation from master cluster to federation?

File: federation/pkg/federation-controller/ingress/ingress_controller.go [hunk 4]
```diff
@@ -624,10 +624,7 @@ func (ic *IngressController) reconcileIngress(ingress types.NamespacedName) {
 			ic.deliverIngress(ingress, 0, true)
 			return
 		}
-		desiredIngress := &extensions_v1beta1.Ingress{
-			ObjectMeta: baseIngress.ObjectMeta,
-			Spec:       baseIngress.Spec,
-		}
+		desiredIngress := &extensions_v1beta1.Ingress{}
 		objMeta, err := conversion.NewCloner().DeepCopy(baseIngress.ObjectMeta)
```
Comment (nikhiljindal) [inline federation/pkg/federation-controller/ingress/ingress_controller.go]: We never use this anymore. We always overwrite desiredIngress.ObjectMeta.
Can delete this code

File: federation/pkg/federation-controller/ingress/ingress_controller.go [hunk 5]
```diff
@@ -707,29 +703,43 @@ func (ic *IngressController) reconcileIngress(ingress types.NamespacedName) {
...
-						"Updating ingress in cluster %s", cluster.Name)
-
-					operations = append(operations, util.FederatedOperation{
-						Type:        util.OperationTypeUpdate,
-						Obj:         desiredIngress,
-						ClusterName: cluster.Name,
-					})
-					// TODO: Transfer any readonly (target-proxy, url-map etc) annotations from the master cluster to the federation, if this is the master cluster.
-					// This is only for consistency, so that the federation ingress metadata matches the underlying clusters.  It's not actually required.
+				objMeta, err := conversion.NewCloner().DeepCopy(clusterIngress.ObjectMeta)
```
Comment (nikhiljindal) [inline federation/pkg/federation-controller/ingress/ingress_controller.go]: optional: I liked the previous code where we did all of this only if ObjectMeta is different.
Dont need to do all of this (recompute ObjectMeta) if ObjectMetaEquivalent is true.

**Comment by nikhiljindal** [comment]
All the comments are optional improvements. Fine to do in a follow up

**Comment by pwittrock** [comment]
@quinton-hoole @nikhiljindal Looks like the tests are failing and no lgtm.  Is the current plan still to cherrypick this tonight?

**Comment by madhusudancs** [comment]
The tests are still not passing, but this PR is definitely fixing a known bug. So I will let this PR merge. Opened issue #33350 and assigned it to @quinton-hoole for follow up.

File: federation/pkg/federation-controller/ingress/ingress_controller.go (hunk 3 above)
Comment (quinton-hoole) [inline federation/pkg/federation-controller/ingress/ingress_controller.go]: The ingress controller in the cluster applies annotations to indicate what target-proxy, url-map etc it applied to this ingress.  It should all be consistent across all cluster ingresses, and should hence also be applied 'upwards' to the federated ingress for consistency.  Not 100% required, but ensures strict API compatibility between Kubernetes and Federation.  Make sense?

Comment (nikhiljindal) [inline federation/pkg/federation-controller/ingress/ingress_controller.go]: yup, it does now!
Thanks :)

**Comment by quinton-hoole** [comment]
@k8s-bot federation gce e2e test this

**Comment by quinton-hoole** [comment]
I found the bug in the e2e tests that was causing them to fail (PR #33372 fixes it).  And I've tested this PR manually and it works. So I think we should cherrypick it in to v1.4. @pwittrock at your convenience.

**Comment by quinton-hoole** [comment]
PS: I'll be online most of the weekend so don't be shy to ping me if you need me to do anything to get the cherrypick in.

File: federation/pkg/federation-controller/ingress/ingress_controller.go (hunk 4 above)
Comment (quinton-hoole) [inline federation/pkg/federation-controller/ingress/ingress_controller.go]: No, we don't always overwrite it.  But I agree, the code is confusing, and in need ot refactoring.  Will do that holistically in a followup PR.

**Pull request merged**